python/ray/data/datasource/datasource.py

```diff
@@ -28,14 +46,201 @@ def get_column_renames(self) -> Optional[Dict[str, str]]:
             A dictionary mapping old column names to new column names,
             or None if no renaming has been applied.
         """
-        return None
+        return self._data_columns_rename_map if self._data_columns_rename_map else None
+
+    @staticmethod
+    def _apply_column_mapping(
+        columns: List[str],
+        mapping: Dict[str, str],
+    ) -> List[str]:
+        """Apply a column name mapping, keeping unmapped columns as-is.
+
+        Args:
+            columns: List of column names to map
+            mapping: Dictionary mapping from one name to another
+
+        Returns:
+            List of column names with mapping applied
+        """
+        return [mapping.get(col, col) for col in columns]
+
+    @staticmethod
+    def _process_projection_with_renames(
+        columns: Optional[List[str]],
+        existing_rename_map: Optional[Dict[str, str]],
+    ) -> _ProjectionProcessingResult:
+        """Process column projection accounting for existing renames.
+
+        When columns have been renamed, this handles:
+        1. Translating column names from renamed space back to original names
+        2. Filtering the rename map to only include selected columns
+
+        Args:
+            columns: Columns to select (in renamed space if rename_map exists)
+            existing_rename_map: Current rename mapping (original -> renamed)
+
+        Returns:
+            _ProjectionProcessingResult containing:
+            - rebound_columns: Column names in original/storage space
+            - filtered_rename_map: Rename map containing only selected columns
+        """
+        if not existing_rename_map or columns is None:
+            # No renames to process
+            return _ProjectionProcessingResult(
+                rebound_columns=columns,
+                filtered_rename_map=existing_rename_map,
+            )
```

Comment:
Let's actually combine projection and rename map into a single dict (so that they couldn't get out of sync)

Reply:
That way combining them becomes simply transitive intersection of 2 maps